src/lex.cpp

```diff
@@ -634,34 +657,55 @@ void lexer::parse_binary_number() {
   this->input_ = check_garbage_in_number_literal(input);
 }
 
-const char8* lexer::check_garbage_in_number_literal(const char8* input) {
+void lexer::parse_octal_number(bool strict) {
+  const char8* input = this->input_;
+
+  while (this->is_octal_digit(*input)) {
+    ++input;
+  }
+
   const char8* garbage_begin = input;
   for (;;) {
     switch (*input) {
+    /* TODO (👮🏾‍♀️ mswitch switch to just 8 and 9)*/
+    // decimal point should also be included
     QLJS_CASE_DECIMAL_DIGIT:
+      if (strict) {
+        ++input;
+        goto done_parsing_garbage;
+      } else {
+        this->input_ = input;
+        this->parse_number();
+        return;
+      }
     QLJS_CASE_IDENTIFIER_START:
-      input += 1;
-      break;
+    case u8'.':
+      ++input;
+      goto done_parsing_garbage;
       default:
         goto done_parsing_garbage;
     }
   }
+  /* note(👮🏾‍♀️)this is bad, because it won't parse the trailing
+   * garbage separate tokens correctly. for example, `0o57.2` should be two
+   * number tokens, while `0o5a` should be an "unknown" token and an "any" token
+   */
 done_parsing_garbage:
   const char8* garbage_end = input;
   if (garbage_end != garbage_begin) {
-    this->error_reporter_->report(error_unexpected_characters_in_number{
+    this->error_reporter_->report(error_unexpected_characters_in_octal_number{
         source_code_span(garbage_begin, garbage_end)});
     input = garbage_end;
   }
-
-  return input;
+  this->input_ = input;
 }
 
 void lexer::parse_number() {
   QLJS_ASSERT(this->is_digit(this->input_[0]) || this->input_[0] == '.');
   const char8* number_begin = this->input_;
   const char8* input = number_begin;
 
+  // @@@ (👮🏾‍♀️)can I use this other places?
```

Comment:
Already done in https://github.com/strager/quick-lint-js/commit/4ecd4c69ec86c0f97b5a1af389796f2fc4693c5e

But unfortunately won't use it in all of my implementations

```suggestion
```